@Nested (Junit5) associated with @QuarkusTest causes an UnsatisfiedResolutionException

**Description**
An `UnsatisfiedResolutionException` will be thrown when a test class is annotated with `@QuarkusTest` and implements inner test classes in connection with `@Nested` (Junit5).

**Actual behavior**
The relevant dump of the stack trace (via `mvn clean verify`):
```
...
org.junit.jupiter.api.extension.TestInstantiationException: Failed to create test instance
Caused by: java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
Caused by: java.lang.reflect.InvocationTargetException
Caused by: javax.enterprise.inject.UnsatisfiedResolutionException: No bean found for required type [class org.acme.testcoverage.GreetingResourceTest$NestedTestClass] and qualifiers [[]]
...
```

**Steps to reproduce the behavior**
1. Checkout https://github.com/quarkusio/quarkus-quickstarts/tree/master/tests-with-coverage-quickstart 
2. Adjust https://github.com/quarkusio/quarkus-quickstarts/blob/master/tests-with-coverage-quickstart/src/test/java/org/acme/testcoverage/GreetingResourceTest.java

```
package org.acme.testcoverage;

import static io.restassured.RestAssured.given;
import static org.hamcrest.CoreMatchers.is;

import java.util.UUID;

import org.junit.jupiter.api.Nested;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;

import io.quarkus.test.junit.QuarkusTest;

@QuarkusTest
@Tag("integration")
public class GreetingResourceTest {

    @Nested
    class NestedTestClass {
        @Test
        public void testHelloEndpoint() {
            given()
                    .when().get("/hello")
                    .then()
                    .statusCode(200)
                    .body(is("hello"));
        }
    }

    @Test
    public void testGreetingEndpoint() {
        String uuid = UUID.randomUUID().toString();
        given()
                .pathParam("name", uuid)
                .when().get("/hello/greeting/{name}")
                .then()
                .statusCode(200)
                .body(is("hello " + uuid));
    }
}
```

3. Execute `mvn clean verify`

**Environment**
 - Output of `java -version`: 
```
openjdk version "11.0.7" 2020-04-14
OpenJDK Runtime Environment GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02)
OpenJDK 64-Bit Server VM GraalVM CE 20.1.0 (build 11.0.7+10-jvmci-20.1-b02, mixed mode, sharing)
```
 - Output of `mvnw --version`: 
```
Apache Maven 3.6.3 (cecedd343002696d0abb50b32b541b8a6ba2883f)
Maven home: /usr/local/Cellar/maven/3.6.3_1/libexec
Java version: 11.0.7, vendor: GraalVM Community, runtime: /Library/Java/JavaVirtualMachines/graalvm-ce-java11-20.1.0/Contents/Home
...
```
 - Quarkus version: 1.6.1.Final

--- a/integration-tests/main/src/test/java/io/quarkus/it/main/JaxbTestCase.java
+++ b/integration-tests/main/src/test/java/io/quarkus/it/main/JaxbTestCase.java
@@ -2,6 +2,7 @@ package io.quarkus.it.main;
 
 import static org.hamcrest.Matchers.contains;
 
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.Test;
 
 import io.quarkus.test.junit.QuarkusTest;
@@ -10,9 +11,13 @@ import io.restassured.RestAssured;
 @QuarkusTest
 public class JaxbTestCase {
 
-    @Test
-    public void testNews() {
-        RestAssured.when().get("/test/jaxb/getnews").then()
-                .body("author", contains("Emmanuel Bernard"));
+    @Nested
+    class SomeClass {
+        @Test
+        public void testNews() {
+            RestAssured.when().get("/test/jaxb/getnews").then()
+                    .body("author", contains("Emmanuel Bernard"));
+        }
     }
+
 }

--- a/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
+++ b/test-framework/junit5/src/main/java/io/quarkus/test/junit/QuarkusTestExtension.java
@@ -13,8 +13,10 @@ import java.nio.file.Files;
 import java.nio.file.Path;
 import java.nio.file.Paths;
 import java.util.AbstractMap;
+import java.util.ArrayDeque;
 import java.util.ArrayList;
 import java.util.Collections;
+import java.util.Deque;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Locale;
@@ -40,6 +42,7 @@ import org.jboss.jandex.FieldInfo;
 import org.jboss.jandex.Index;
 import org.jboss.jandex.Type;
 import org.jboss.logging.Logger;
+import org.junit.jupiter.api.Nested;
 import org.junit.jupiter.api.extension.AfterAllCallback;
 import org.junit.jupiter.api.extension.AfterEachCallback;
 import org.junit.jupiter.api.extension.BeforeAllCallback;
@@ -119,6 +122,8 @@ public class QuarkusTestExtension
     private static List<Function<Class<?>, String>> testHttpEndpointProviders;
 
     private static DeepClone deepClone;
+    //needed for @Nested
+    private static Deque<Class<?>> currentTestClassStack = new ArrayDeque<>();
 
     private ExtensionState doJavaStart(ExtensionContext context, Class<? extends QuarkusTestProfile> profile) throws Throwable {
         quarkusTestProfile = profile;
@@ -382,7 +387,7 @@ public class QuarkusTestExtension
 
     @Override
     public void beforeEach(ExtensionContext context) throws Exception {
-        if (isNativeTest(context)) {
+        if (isNativeTest()) {
             return;
         }
         if (!failedBoot) {
@@ -450,7 +455,7 @@ public class QuarkusTestExtension
 
     @Override
     public void afterEach(ExtensionContext context) throws Exception {
-        if (isNativeTest(context)) {
+        if (isNativeTest()) {
             return;
         }
         if (!failedBoot) {
@@ -513,8 +518,13 @@ public class QuarkusTestExtension
                 constructor.newInstance(actualTestInstance, actualTestMethod));
     }
 
-    private boolean isNativeTest(ExtensionContext context) {
-        return context.getRequiredTestClass().isAnnotationPresent(NativeImageTest.class);
+    private boolean isNativeTest() {
+        for (Class<?> i : currentTestClassStack) {
+            if (i.isAnnotationPresent(NativeImageTest.class)) {
+                return true;
+            }
+        }
+        return false;
     }
 
     private ExtensionState ensureStarted(ExtensionContext extensionContext) {
@@ -569,9 +579,10 @@ public class QuarkusTestExtension
 
     @Override
     public void beforeAll(ExtensionContext context) throws Exception {
+        currentTestClassStack.push(context.getRequiredTestClass());
         //set the right launch mode in the outer CL, used by the HTTP host config source
         ProfileManager.setLaunchMode(LaunchMode.TEST);
-        if (isNativeTest(context)) {
+        if (isNativeTest()) {
             return;
         }
         ensureStarted(context);
@@ -609,7 +620,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptBeforeAllMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -625,7 +636,7 @@ public class QuarkusTestExtension
     @Override
     public <T> T interceptTestClassConstructor(Invocation<T> invocation,
             ReflectiveInvocationContext<Constructor<T>> invocationContext, ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             return invocation.proceed();
         }
         ExtensionState state = ensureStarted(extensionContext);
@@ -672,8 +683,15 @@ public class QuarkusTestExtension
             Class<?> previousActualTestClass = actualTestClass;
             actualTestClass = Class.forName(extensionContext.getRequiredTestClass().getName(), true,
                     Thread.currentThread().getContextClassLoader());
-
-            actualTestInstance = runningQuarkusApplication.instance(actualTestClass);
+            if (extensionContext.getRequiredTestClass().isAnnotationPresent(Nested.class)) {
+                Class<?> parent = actualTestClass.getEnclosingClass();
+                Object parentInstance = runningQuarkusApplication.instance(parent);
+                Constructor<?> declaredConstructor = actualTestClass.getDeclaredConstructor(parent);
+                declaredConstructor.setAccessible(true);
+                actualTestInstance = declaredConstructor.newInstance(parentInstance);
+            } else {
+                actualTestInstance = runningQuarkusApplication.instance(actualTestClass);
+            }
 
             Class<?> resM = Thread.currentThread().getContextClassLoader().loadClass(TestHTTPResourceManager.class.getName());
             resM.getDeclaredMethod("inject", Object.class, List.class).invoke(null, actualTestInstance,
@@ -696,7 +714,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptBeforeEachMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -707,7 +725,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptTestMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -718,7 +736,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptTestTemplateMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -730,7 +748,7 @@ public class QuarkusTestExtension
     @Override
     public <T> T interceptTestFactoryMethod(Invocation<T> invocation,
             ReflectiveInvocationContext<Method> invocationContext, ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             return invocation.proceed();
         }
         T result = (T) runExtensionMethod(invocationContext, extensionContext);
@@ -741,7 +759,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptAfterEachMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -752,7 +770,7 @@ public class QuarkusTestExtension
     @Override
     public void interceptAfterAllMethod(Invocation<Void> invocation, ReflectiveInvocationContext<Method> invocationContext,
             ExtensionContext extensionContext) throws Throwable {
-        if (isNativeTest(extensionContext)) {
+        if (isNativeTest()) {
             invocation.proceed();
             return;
         }
@@ -816,11 +834,15 @@ public class QuarkusTestExtension
 
     @Override
     public void afterAll(ExtensionContext context) throws Exception {
-        if (!isNativeTest(context) && (runningQuarkusApplication != null)) {
-            popMockContext();
-        }
-        if (originalCl != null) {
-            setCCL(originalCl);
+        try {
+            if (!isNativeTest() && (runningQuarkusApplication != null)) {
+                popMockContext();
+            }
+            if (originalCl != null) {
+                setCCL(originalCl);
+            }
+        } finally {
+            currentTestClassStack.pop();
         }
     }
 

